include/libpmemobj++/experimental/mpsc_queue.hpp

```diff
@@ -43,22 +47,57 @@ class mpsc_queue {
 		size_t len;
 
 		struct iterator {
-			iterator(char *data) : data(data)
+			iterator(char *data, char *end) : data(data), end(end)
 			{
+				/* Advance to first, unconsumed element */
+				while (this->data < end &&
+				       reinterpret_cast<first_block *>(
```

Comment:
i'm not the biggest fan of accessing fields from a casted variable. but I'll leave it up to you.